overlord/servicestate/quota_handlers.go

```diff
@@ -123,37 +124,92 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
 		// in state the fact that the changes were made,
 		// to avoid repeating them.
 		// What remains for this task handler is just to
-		// restart services which will happen regardless if we
-		// get rebooted after unlocking the state - if we got
-		// rebooted before unlocking the state, none of the
-		// changes we made to state would be persisted and we
-		// would run through everything above here again, but
-		// the second time around EnsureSnapServices would end
-		// up doing nothing since it is idempotent.  So in the
-		// rare case that snapd gets restarted but is not a
+		// refresh security profiles and restart services which
+		// will happen regardless if we get rebooted after
+		// unlocking the state - if we got rebooted before unlocking
+		// the state, none of the changes we made to state would
+		// be persisted and we would run through everything above
+		// here again, but the second time around EnsureSnapServices
+		// would end up doing nothing since it is idempotent.
+		// So in the rare case that snapd gets restarted but is not a
 		// reboot also record which services do need
 		// restarting. There is a small chance that services
 		// will be restarted again but is preferable to the
 		// quota not applying to them.
-		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+		if err := rememberQuotaStateUpdated(t, servicesAffected, refreshProfiles); err != nil {
 			return err
 		}
+	}
+
+	if len(servicesAffected) > 0 {
+		prevTask := t
+		if refreshProfiles {
+			prevTask = addRefreshProfileTasks(st, prevTask, servicesAffected)
+		}
+		if err := addRestartServicesTask(st, prevTask, qc.QuotaName, servicesAffected); err != nil {
+			return err
+		}
+	}
+
+	t.SetStatus(state.DoneStatus)
+	return nil
+}
+
+func addRefreshProfileTasks(st *state.State, t *state.Task, servicesAffected map[*snap.Info][]*snap.AppInfo) *state.Task {
+	chg := t.Change()
+	prevTask := t
+	for info := range servicesAffected {
+		setupProfilesTask := st.NewTask("setup-profiles", fmt.Sprintf(i18n.G("Update snap %q (%s) security profiles"), info.SnapName(), info.Revision))
+		setupProfilesTask.Set("snap-setup", &snapstate.SnapSetup{
+			SideInfo: &snap.SideInfo{
+				RealName: info.SnapName(),
+				Revision: info.Revision,
+			},
+		})
+		setupProfilesTask.WaitFor(prevTask)
+		chg.AddTask(setupProfilesTask)
+		prevTask = setupProfilesTask
+	}
+	return prevTask
+}
+
+func addRestartServicesTask(st *state.State, t *state.Task, grpName string, servicesAffected map[*snap.Info][]*snap.AppInfo) error {
+	chg := t.Change()
+	restartTask := st.NewTask("quota-restart-services", fmt.Sprintf("Restarting services for quota %q", grpName))
+	if err := rememberQuotaStateUpdated(restartTask, servicesAffected, false); err != nil {
+		return err
+	}
+	restartTask.WaitFor(t)
+	chg.AddTask(restartTask)
+	return nil
+}
+
+func (m *ServiceManager) doQuotaRestartServices(t *state.Task, _ *tomb.Tomb) error {
```

Comment:
I recommend not introducing a new task (also because it would need new bits of conflict logic), instead I would do some of the computations done in restartSnapServices upfront and create a service-control task instead with the right action

Reply:
Ah that is neat, I did not realize that we already had a task for doing restarting services. Thanks for pointing that out. I'll refactor this a bit.